PWGUD/Core/UPCJpsiCentralBarrelCorrHelper.h

```diff
@@ -0,0 +1,193 @@
+{
+  TLorentzVector pa(1., 0., 0, 1);  // projectile
+  TLorentzVector pb(1., 0., 0, -1); // target
+
+  float* q = new float[3]; // to save values
+
+  // Accoplanarity angle
+  Float_t deltaPhi;
+  deltaPhi = lv1->Phi() - lv2->Phi();
+  float accoplCut = 1. - abs(deltaPhi) / o2::constants::math::PI;
+  // z
+  TLorentzVector z;
+  Float_t part1, part2;
+
+  // Dot product: v1*v2 = t1*t2-x1*x2-y1*y2-z1*z2
+
+  part1 = lv->Dot(pb);
+  part2 = lv->Dot(pa);
+
+  Float_t part3x = pa.X() * part1;
+  Float_t part3y = pa.Y() * part1;
+  Float_t part3z = pa.Z() * part1;
+  Float_t part3e = pa.T() * part1;
+
+  Float_t part4x = pb.X() * part2;
+  Float_t part4y = pb.Y() * part2;
+  Float_t part4z = pb.Z() * part2;
+  Float_t part4e = pb.T() * part2;
+
+  TLorentzVector part3(TVector3(part3x, part3y, part3z), part3e);
+  TLorentzVector part4(TVector3(part4x, part4y, part4z), part4e);
+
+  // Un-normalized Z
+  z = part3 - part4;
+
+  // Normalized z
+  Float_t normz = TMath::Sqrt(-z * z);
+  Float_t znx = z.X() / normz;
+  Float_t zny = z.Y() / normz;
+  Float_t znz = z.Z() / normz;
+  Float_t zne = z.E() / normz;
+
+  // Normalized z
+  TLorentzVector zhat(TVector3(znx, zny, znz), zne);
+
+  // calculate x
+  TLorentzVector x;
+
+  Float_t constant1 = (lv->Dot(*lv)) / (2 * (lv->Dot(pa)));
+  Float_t constant2 = (lv->Dot(*lv)) / (2 * (lv->Dot(pb)));
+
+  Float_t comp1x = pa.X() * constant1;
+  Float_t comp1y = pa.Y() * constant1;
+  Float_t comp1z = pa.Z() * constant1;
+  Float_t comp1e = pa.T() * constant1;
+
+  TLorentzVector comp1(TVector3(comp1x, comp1y, comp1z), comp1e);
+
+  Float_t comp2x = pb.X() * constant2;
+  Float_t comp2y = pb.Y() * constant2;
+  Float_t comp2z = pb.Z() * constant2;
+  Float_t comp2e = pb.T() * constant2;
+
+  TLorentzVector comp2(TVector3(comp2x, comp2y, comp2z), comp2e);
+
+  // Un-normalized x
+  x = *lv - comp1 - comp2;
+  // normalize x
+  Float_t normx = TMath::Sqrt(-x * x);
+  Float_t xnx = x.X() / normx;
+  Float_t xny = x.Y() / normx;
+  Float_t xnz = x.Z() / normx;
+  Float_t xne = x.E() / normx;
+
+  // Normalized x
+  TLorentzVector xhat(TVector3(xnx, xny, xnz), xne);
+
+  // calculate y
+  // TLorentzVector y;
+  Float_t yone = pa.Y() * pb.Z() * lv->E() - pa.Z() * pb.Y() * lv->E() + pa.Z() * pb.E() * lv->Y() + pa.E() * pb.Y() * lv->Z() - pa.Y() * pb.E() * lv->Z() - pa.E() * pb.Z() * lv->Y();
+  Float_t ytwo = -pa.Z() * pb.E() * lv->X() + pa.Z() * pb.X() * lv->E() - pa.X() * pb.Z() * lv->E() + pa.X() * pb.E() * lv->Z() - pa.E() * pb.X() * lv->Z() + pa.E() * pb.Z() * lv->X();
+  Float_t ythree = pa.X() * pb.Y() * lv->E() - pa.Y() * pb.X() * lv->E() + pa.Y() * pb.E() * lv->X() - pa.X() * pb.E() * lv->Y() + pa.E() * pb.X() * lv->Y() - pa.E() * pb.Y() * lv->X();
+  Float_t yfour = -pa.X() * pb.Y() * lv->Z() + pa.X() * pb.Z() * lv->Y() - pa.Z() * pb.X() * lv->Y() + pa.Z() * pb.Y() * lv->X() - pa.Y() * pb.Z() * lv->X() + pa.Y() * pb.X() * lv->Z();
```

Comment:
I admit I have no idea what this does, but I guess it might help to have these calculations defined in some utility functions with a proper documentation.

Reply:
What would be the benefit of having it in a utility function?

There calculations are based on a specific approach and may choose with advancing analysis. They are solely for my use.